### PR TITLE
[JavaToolInstallerV0] - add support of bin in the root, JDK file validation

### LINKED ISSUE #12858: [DTS 1674974] Bug in Java tool installer task with JDK Source option Local Directory
Please look at DTS 1674974 that has been opened here:
https://dev.azure.com/mseng/_workitems/edit/1674974

### PR DESCRIPTION
**Pull request by AlexandrKravchuk**
Closes #12858 
- [x] Added support of 'bin' folder right inside the JDK archive
- [x] Added error handler which informs the user if the JDK Archive is valid

### CODE REVIEW

File: Tasks/JavaToolInstallerV0/task.json [hunk 1]
```diff
@@ -171,6 +171,7 @@
         "CouldNotFetchAccessTokenforMSIDueToMSINotConfiguredProperlyStatusCode": "Could not fetch access token for Managed Service Principal. Please configure Managed Service Identity (MSI) for virtual machine 'https://aka.ms/azure-msi-docs'. Status code: %s, status message: %s",
         "CouldNotFetchAccessTokenforMSIStatusCode": "Could not fetch access token for Managed Service Principal. Status code: %s, status message: %s",
         "ExpiredServicePrincipal": "Could not fetch access token for Azure. Verify if the Service Principal used is valid and not expired.",
-        "CorrelationIdForARM": "Correlation ID from ARM api call response : %s"
+        "CorrelationIdForARM": "Correlation ID from ARM api call response : %s",
+        "WrongArchiveStructure": "JDK file is not valid. Verify if JDK file contains only one root folder with 'bin' inside."
```
[the remark this answers is not in the document]
Comment (AlexandrKravchuk) [reply Tasks/JavaToolInstallerV0/task.json:175]: JDK file is the name of the field. This is how it's mentioned in the docs. So I think it is quite clear.
Screenshot 2020-05-19 at 18 54 24

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 2]
```diff
@@ -131,6 +133,25 @@ export class JavaFilesExtractor {
...
     }
 
+    /**
+     * Creates a list of directories on the root level of structure.
+     * @param pathsArray - contains paths to all the files inside the structure
+     * @param root - path to the directory we want to get the structure of
+     */
+
+    private static sliceStructure(pathsArray: Array<string>, root: string = pathsArray[0]): Array<string>{
+        const dirPathLength = root.length;
```
[the remark this answers is not in the document]
Comment (AlexandrKravchuk) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:143]: Typing constants in TypeScript is redundant. Besides this is a code style that is used throughout the whole project.

[the remark this answers is not in the document]
Comment (AlexandrKravchuk) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:143]: Typing constants in TypeScript is redundant. Besides this is a code style that is used throughout the whole project.

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 3]
```diff
@@ -131,6 +133,25 @@ export class JavaFilesExtractor {
         }    
     }
 
+    /**
+     * Creates a list of directories on the root level of structure.
+     * @param pathsArray - contains paths to all the files inside the structure
+     * @param root - path to the directory we want to get the structure of
+     */
+
```
Comment (anatolybolshakov) [inline Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:141]: Nit: looks like redundant string

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 4]
```diff
@@ -131,6 +133,25 @@ export class JavaFilesExtractor {
...
+    private static sliceStructure(pathsArray: Array<string>, root: string = pathsArray[0]): Array<string>{
+        const dirPathLength = root.length;
+        const structureObject = {};
```
Comment (anatolybolshakov) [inline Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:144]: Could you add type to this dictionary? It's unclear from init value here which type key/value have

**Comment by AlexandrKravchuk** [comment]
Run for archive with 'bin' inside:
Screenshot 2020-05-19 at 19 53 41
Run for archive without 'bin' inside:
Screenshot 2020-05-19 at 19 53 56
Run for archive with 'bin' inside 'jdk-8u202-windows-x64':
Screenshot 2020-05-19 at 19 54 39

File: Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts [hunk 5]
```diff
@@ -148,9 +172,19 @@ export class JavaFilesExtractor {
         const jdkFile = path.normalize(repoRoot);
         const stats = taskLib.stats(jdkFile);
         if (stats.isFile()) {
-            await this.extractFiles(jdkFile, fileEnding)
+            await this.extractFiles(jdkFile, fileEnding);
             finalDirectoriesList = taskLib.find(this.destinationFolder).filter(x => taskLib.stats(x).isDirectory());
-            jdkDirectory = finalDirectoriesList.filter(dir => initialDirectoriesList.indexOf(dir) < 0)[0];
+            const freshExtractedDirectories = finalDirectoriesList.filter(dir => initialDirectoriesList.indexOf(dir) < 0);
```
Comment (leantk) [inline Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:177]: Can we add some more comments explaining what this does a little better please? I see that we are looking at the extracted directories looking for the bin directory, but then I'm a little confused by the if/else logic below

Comment (AlexandrKravchuk) [reply Tasks/JavaToolInstallerV0/FileExtractor/JavaFilesExtractor.ts:177]: Fair enough. Usage of De Morgans's law made this expression a little complicated.
I tried to refactor this part to make it easier to understand.

**Pull request merged**